tantar/src/main/java/org/opencadc/inventory/db/DAOConfigurationManager.java

```diff
@@ -0,0 +1,194 @@
+ *  Council of Canada nor the            Recherches du Canada ni les noms
+ *  names of its contributors may        de ses  participants ne peuvent
+ *  be used to endorse or promote        être utilisés pour approuver ou
+ *  products derived from this           promouvoir les produits dérivés
+ *  software without specific prior      de ce logiciel sans autorisation
+ *  written permission.                  préalable et particulière
+ *                                       par écrit.
+ *
+ *  This file is part of the             Ce fichier fait partie du projet
+ *  OpenCADC project.                    OpenCADC.
+ *
+ *  OpenCADC is free software:           OpenCADC est un logiciel libre ;
+ *  you can redistribute it and/or       vous pouvez le redistribuer ou le
+ *  modify it under the terms of         modifier suivant les termes de
+ *  the GNU Affero General Public        la “GNU Affero General Public
+ *  License as published by the          License” telle que publiée
+ *  Free Software Foundation,            par la Free Software Foundation
+ *  either version 3 of the              : soit la version 3 de cette
+ *  License, or (at your option)         licence, soit (à votre gré)
+ *  any later version.                   toute version ultérieure.
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.inventory.db;
+
+import ca.nrc.cadc.db.ConnectionConfig;
+import ca.nrc.cadc.db.DBUtil;
+import ca.nrc.cadc.util.StringUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import javax.naming.NamingException;
+
+import org.opencadc.inventory.Entity;
+import org.opencadc.inventory.InventoryUtil;
+
+
+/**
+ * Class to manage and create DAO instances.  This will create new DAOs, or build one from an existing DAO to have
+ * true transaction management.
```

Comment:
The `true` was meant to illustrate that the Transaction Management was correct in that the DAO that created the transaction needs to be passed into new DAOs to maintain the transaction.

Reply:
The copy constructor in a DAO makes it share config and DataSource; that, plus the thread-bound transactions in spring jdbc cause statements to be executed in a transaction. The standard way to do it is to create one DAO directly and create others locally as necessary using the copy constructor.

The minoc PutAction (for example) has an ArtifactDAO that it uses for manage the txn(s) and instantiates other DAO classes as needed (eg starting at line ~206).

The subtle thing here is that tantar needs two database connections (one for Iterator<Artifact> and one for ArtifactDAO + event DAO operations during resolve). That way the resolve txns are not part of the open (autocommit) read txn... not exactly sure what happens there because I never tried it, but presumably if you have only one connection then those resolve txns are actually nested txns within the read... I can see how a single connection would appear to work fine for an inital load (because the Iterator<Artifact> reaches the end and the work all happens in the simple loop over Iterator<StorageMetadata>.

So, in BucketValidator I would create two single connection DataSource(s), bind in JNDI context, and then create two ArtifactDAO instances, eg:
```protected final ArtifactDAO iteratorDAO;
protected final ArtifactDAO artifactDAO;```
Use the first one to iterate and the second (to manage txns, directly, and via copy constructor for events) for resolve transactions. The handlful of lines of code all in one place to instantiate each of those makes the database setup really clear and understandable.

Reply:
Done.